base/sort.jl

```diff
@@ -1109,12 +1304,114 @@ function sort!(A::AbstractArray;
     A
 end
 
+## sorting serialization to alow radix sorting primitives other than UInts ##
+module Serial
+using ...Order
+using ..Base: @inbounds, @eval, AbstractVector, nothing, signed, unsigned,
+    typemin, xor, reinterpret, Signed, Unsigned, Type, eltype
+
+"""
+    Serializable(T::Type, order::Ordering)
+
+Return `typeof(serialize(x::T, order))` if [`serialize`](@ref) and
+[`deserialize`](@ref) are implemented.
+
+If either is not implemented, return nothing.
+"""
+Serializable(T::Type, order::Ordering) = nothing
+
+"""
+    serialize(x, order::Ordering)::Unsigned
+
+Map `x` to an un unsigned integer, maintaining sort order.
+
+The map should be reversible with [`deserialize`](@ref), so `isless(order, a, b)` must be
+a linear ordering for `a, b <: typeof(x)`. Satisfies
+`isless(order, a, b) === (serialize(order, a) < serialize(order, b))`
+and `x === deserialize(typeof(x), serialize(order, x), order)`
+
+See also: [`Serializable`](@ref) [`deserialize`](@ref)
+"""
+function serialize end
+
+"""
+    deserialize(T::Type, u::Unsigned, order::Ordering)
+
+Reconstruct the unique value `x::T` that serializes to `u`. Satisfies
+`x === deserialize(T, order, serialize(order, x::T))` for all `x <: T`.
+
+See also: [`serialize`](@ref) [`Serializable`](@ref)
+"""
+function deserialize end
+
+
+### Primitive Types
+
+# Integers
+serialize(x::Unsigned, ::ForwardOrdering) = x
+deserialize(::Type{T}, u::T, ::ForwardOrdering) where T <: Unsigned = u
+
+serialize(x::Signed, ::ForwardOrdering) =
+    unsigned(xor(x, typemin(x)))
+deserialize(::Type{T}, u::Unsigned, ::ForwardOrdering) where T <: Signed =
+    xor(signed(u), typemin(T))
+
+# unsigned(Int) is not available during bootstrapping.
+for (U, S) in [(UInt8, Int8), (UInt16, Int16), (UInt32, Int32), (UInt64, Int64), (UInt128, Int128)]
+    @eval Serializable(::Type{<:Union{$U, $S}}, ::ForwardOrdering) = $U
+end
+
+# Floats are not Serializable under regular orderings because they fail on NaN edge cases.
+# Float serialization is defined in ..Float, where the Left and Right orderings guarantee
+# that there are no NaN values
+
+# Booleans
+# serialize! fails on AbstractVector{Bool} because the reinterpret is sketchy. try:
+#     u = reinterpret(UInt8, trues(10)); u[3] = 0x17; print(u)
```

Comment:
What's the problem exactly with `0x17` being converted silently to `0x01`? Isn't it enough to map `false` to `0x00` and `true` to `0x01`?

Reply:
`sort!(collect(trues(501)), alg=Sort.AdaptiveSort(QuickSort), rev=true)` used to throw an error because `false` went to `0xff` and `true` went to `0xfe`. Apparently, that's not an issue anymore. We still never use serialize on Bool in Base because we count sort instead. We should either delete Bool entirely without comment (bc it isn't used) or leave it in (for completeness). I lean toward the former.